File: languages/c/src/benchmarker/benchmarker.c
#include "benchmarker.h"

#define NANO_PER_MILLI 1000000L

// Prints execution time, converting nanoseconds into milliseconds up to 4 decimal places
void print_benchmark_analysis(char* operation_name, int execution_count, double time_in_nanoseconds) {
    printf("C's %s execution time (over %d loops): %.6f ms\n", 
        operation_name,
        execution_count,
        time_in_nanoseconds / NANO_PER_MILLI
    );
}

void execute_benchmarker(BenchmarkerContext *context) {
    if (context->benchmarker != NULL && context->benchmarker != NULL && context->benchmarker->execute != NULL) {
        context->benchmarker->execute(context);
    }
}

void clean_up_benchmarker(BenchmarkerContext *context) {
    if (context->benchmarker != NULL && context->benchmarker->clean_up != NULL) {
        context->benchmarker->clean_up(context);
    }
}

File: languages/zig/src/benchmarkers/benchmarker.zig
// Contains generic benchmarking logic

const std = @import("std");

pub fn Benchmarker(comptime T: type) type {
    return struct {
        operation_name: []const u8,

        context: T,

        getOperationExecutionTime: *const fn (self: Self) anyerror!u64,

        cleanUp: *const fn (self: Self) void,

        const Self = @This();

        fn getOperationExecutionResults(self: Self, execution_count: usize) anyerror!std.ArrayList(u64) {
            const allocator = std.heap.page_allocator;
            var result_arr = std.ArrayList(u64).init(allocator);
            errdefer result_arr.deinit();

            for (0..execution_count) |_| {
                const execution_time = try self.getOperationExecutionTime(self);
                try result_arr.append(execution_time);
            }

            return result_arr;
        }

        pub fn printBenchmarkAnalysis(self: Self, execution_count: usize) void {
            const results = self.getOperationExecutionResults(execution_count) catch |err| {
                std.debug.panic("An error occurred while attempting to get execution results: {}", .{err});
            };
            defer results.deinit();

            // TODO: Print other values besides total execution time
            var total_nano_time: f64 = 0;
            for (0..execution_count) |i| {
                total_nano_time += @floatFromInt(results.items[i]);
            }

            const stdout = std.io.getStdOut().writer();
            stdout.print("Zig's {s} execution time (over {d} loops): {d:.6} ms\n", .{ self.operation_name, execution_count, total_nano_time / 1000000 }) catch unreachable;
        }
    };
}
